lib/Onnxifi/Base.cpp

```diff
@@ -48,22 +48,80 @@ onnxStatus Graph::initGraph(const void *onnxModel, size_t onnxModelSize,
   // TODO: support multiple functions here.
   function_ = backendPtr_->getEE().getModule().createFunction("inference");
 
-  std::unique_ptr<ModelLoader> loader = ModelLoader::parse(
-      onnxModel, onnxModelSize, weightCount, weightDescriptors, *function_);
+  // Keep loader to have an access to the input and output ONNX vars.
+  // TODO: The loader itself is not strictly required, remove that
+  // and just keep mapping.
+  loader_ = ModelLoader::parse(onnxModel, onnxModelSize, weightCount,
+                               weightDescriptors, *function_);
   // TODO: make better error reporting.
-  if (!loader) {
+  if (!loader_) {
     return ONNXIFI_STATUS_INTERNAL_ERROR;
   }
 
+  // Emit IR for the graph and compile it.
+  backendPtr_->getEE().compile(CompilationMode::Infer, function_);
+
   return ONNXIFI_STATUS_SUCCESS;
 }
 
-onnxStatus Graph::run() { return ONNXIFI_STATUS_SUCCESS; }
+onnxStatus Graph::run() {
+  // Copy tensors from the input addresses to the Glow tensors.
+  std::vector<Tensor *> tensors;
+  std::vector<glow::Variable *> vars;
+  for (auto inputVar : inputs_) {
+    auto *var = inputVar.first;
+    auto *type = var->getType();
+    void *inputBuffer = reinterpret_cast<void *>(inputVar.second);
+    tensors.push_back(new Tensor(inputBuffer, type));
+    vars.push_back(var);
+  }
+
+  // Run inference.
+  backendPtr_->getEE().run(vars, tensors);
+
+  // Copy outputs to the addresses specified in the outputs_.
+  for (auto outputVar : outputs_) {
+    void *outputAddress = reinterpret_cast<void *>(outputVar.second);
+    Tensor &res = outputVar.first->getVariable()->getPayload();
+
+    memcpy(outputAddress, res.getUnsafePtr(),
+           res.size() * res.getType().getElementSize());
+  }
+
+  return ONNXIFI_STATUS_SUCCESS;
+}
 
 onnxStatus Graph::setIO(uint32_t inputsCount,
                         const onnxTensorDescriptorV1 *inputDescriptors,
                         uint32_t outputsCount,
                         const onnxTensorDescriptorV1 *outputDescriptors) {
+  // Process inputs.
+  for (unsigned i = 0; i < inputsCount; ++i) {
```

Comment:
@yinghai While running the "conv" ONNXIFI test I noticed that both W and X are passed from the C2 through the ONNXIFI on setGraphIO call. I suppose only X should be passed. (W is a static weights which was initialized as part of the initGraph).

Reply:
Yeah, that's not right. Let me check.

Reply:
OK. It doesn't break things but I'll figure out a way to prune this.

Reply:
It does not break because I had to explicitly filter out weights from here ;) But yes, it needs to be fixed, not urgent.

Reply:
It might be an artifact of the test: https://github.com/pytorch/pytorch/blob/45e5c17ecfbe00708e2c61488d688b3d651ab8ca/caffe2/python/onnx/test_onnxifi.py#L91